src/ssa_ir/ssa_type.rs

```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+
+use solang_parser::pt::Identifier;
+
+use crate::pt::Loc;
+use crate::sema::ast;
+use crate::sema::ast::ArrayLength;
+use crate::sema::ast::ParameterAnnotation;
+use crate::ssa_ir::expr::Operand;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum StructType {
+    UserDefined(usize),
+    SolAccountInfo,
+    SolAccountMeta,
+    SolParameters,
+    ExternalFunction,
+    Vector(Box<Type>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Type {
```

Comment:
So how are we supposed to generate debug information with all the high-level information removed? e.g. user types, enums, strings, etc

Reply:
The idea of lowering the types was to facilitate future optimizations and making the new IR closer to machine level.

Debug information is important, though. What do you suggest here? Keeping a reference to the codegen type?

Reply:
Holding a reference to the codegen type is a good solution.

I think the best solution would be that we have a codegen `Type` which holds a reference to the ast type.

Reply:
> Holding a reference to the codegen type is a good solution.
> 
> I think the best solution would be that we have a codegen `Type` which holds a reference to the ast type.

The codegen type is the same as the AST type.